Src/LiquidProjections.NHibernate/NHibernateEventMapConfigurator.cs

```diff
@@ -36,117 +36,80 @@ internal sealed class NHibernateEventMapConfigurator<TProjection, TKey>
-            Func<TProjection, Task> projector, ProjectionModificationOptions options)
+        private async Task OnCreate(TKey key, NHibernateProjectionContext context, Func<TProjection, Task> projector, Func<TProjection, bool> shouldOverwrite)
         {
             TProjection projection = await cache.Get(key, () => Task.FromResult(context.Session.Get<TProjection>(key)));
-            if (projection == null)
+            if ((projection == null) || shouldOverwrite(projection))
             {
-                switch (options.MissingProjectionBehavior)
+                if (projection == null)
+                {
+                    projection = new TProjection();
+                    setIdentity(projection, key);
+
+                    context.Session.Save(projection);
+                    cache.Add(projection);
+                }
+                else
                 {
-                    case MissingProjectionModificationBehavior.Create:
-                    {
-                        projection = new TProjection();
-                        setIdentity(projection, key);
-
-                        await projector(projection).ConfigureAwait(false);
-                        context.Session.Save(projection);
-                        cache.Add(projection);
-                        break;
-                    }
-
-                    case MissingProjectionModificationBehavior.Ignore:
-                    {
-                        break;
-                    }
-
-                    case MissingProjectionModificationBehavior.Throw:
-                    {
-                        throw new ProjectionException(
-                            $"Projection {typeof(TProjection)} with key {key} does not exist.");
-                    }
-
-                    default:
-                    {
-                        throw new NotSupportedException(
-                            $"Not supported missing projection behavior {options.MissingProjectionBehavior}.");
-                    }
+                    // Reattach it to the session
+                    // See also https://stackoverflow.com/questions/2932716/nhibernate-correct-way-to-reattach-cached-entity-to-different-session
+                    context.Session.Lock(projection, LockMode.None);
                 }
+                
+                await projector(projection).ConfigureAwait(false);
+            }
+        }
+
+        private async Task OnUpdate(TKey key, NHibernateProjectionContext context, Func<TProjection, Task> projector, Func<bool> createIfMissing)
+        {
+            TProjection projection = await cache.Get(key, () => Task.FromResult(context.Session.Get<TProjection>(key)));
+            if ((projection == null) && createIfMissing())
+            {
+                projection = new TProjection();
+                setIdentity(projection, key);
+
+                context.Session.Save(projection);
+                cache.Add(projection);

```

Comment:
Should we create a missing projection when it is filtered out?

Reply:
There's no point in filtering out an object that is just created. In fact, if somebody decides to map an event as `AsUpdateOf().CreateIfMissing`, I don't think they want to filter that projection out.

Reply:
As discussed, I agree 👍